tests/unit/test_charm.py

```diff
@@ -987,3 +994,302 @@ def test_given_ca_cert_is_not_stored_when_certificate_transfer_relation_joins_th
+        patch_get_s3_connection_info,
+    ):
+        mock_session = Mock()
+        mock_resource = Mock()
+        mock_bucket = Mock()
+        mock_client = Mock()
+
+        patch_session.return_value = mock_session
+        mock_session.resource.return_value = mock_resource
+        mock_resource.Bucket.return_value = mock_bucket
+        mock_bucket.meta.client = mock_client
+        mock_bucket.create.side_effect = CustomBotoCoreError()
+        mock_client.head_bucket.side_effect = CustomBotoCoreError()
+        patch_get_s3_connection_info.return_value = {
+            "bucket": "whatever bucket",
+            "access-key": "whatever access key",
+            "secret-key": "whatever secret key",
+            "endpoint": "whatever endpoint",
+            "region": "whatever region",
+        }
+        event = Mock()
+        self.harness.set_leader(is_leader=True)
+        self.harness.add_relation(relation_name=S3_RELATION_NAME, remote_app="s3-integrator")
+        self.harness.charm._on_create_backup_action(event)
+        event.fail.assert_called_with(message="Failed to create S3 bucket.")
+
+    @patch("boto3.session.Session", new=Mock)
+    @patch("vault.Vault.is_initialized")
+    @patch("vault.Vault.is_api_available")
+    @patch(f"{S3_LIB_PATH}.S3Requirer.get_s3_connection_info")
+    def test_given_snapshot_creation_fails_when_create_backup_action_then_action_fails(
+        self,
+        patch_get_s3_connection_info,
+        patch_is_api_available,
+        patch_is_initialized,
+    ):
+        patch_is_initialized.return_value = True
+        patch_is_api_available.return_value = True
+        self.harness.set_can_connect(container=self.container_name, val=True)
+        self.harness.add_storage(storage_name="certs", attach=True)
+        peer_relation_id = self._set_peer_relation()
+        self._set_ca_certificate_secret_in_peer_relation(
+            certificate="whatever certificate",
+            private_key="whatever private key",
+            relation_id=peer_relation_id,
+        )
+        patch_get_s3_connection_info.return_value = {
+            "bucket": "whatever bucket",
+            "access-key": "whatever access key",
+            "secret-key": "whatever secret key",
+            "endpoint": "whatever endpoint",
+            "region": "whatever region",
+        }
+        event = Mock()
+        self.harness.set_leader(is_leader=True)
+        self.harness.add_relation(relation_name=S3_RELATION_NAME, remote_app="s3-integrator")
+        self.harness.charm._on_create_backup_action(event)
+        event.fail.assert_called_with(message="Failed to create raft snapshot.")
+
+    @patch("vault.Vault.is_sealed", new=Mock)
+    @patch("vault.Vault.unseal", new=Mock)
+    @patch("vault.Vault.create_snapshot", new=Mock)
+    @patch("vault.Vault.is_initialized")
+    @patch("vault.Vault.is_api_available")
+    @patch(f"{S3_LIB_PATH}.S3Requirer.get_s3_connection_info")
+    @patch("boto3.session.Session")
```

Comment:
We shouldn't have to mock this anymore since we are not using it directly in the charm code. This applies in other tests as well

Reply:
Keeping the mock of this external dependency would make the tests faster.
However I removed it and I am following the same pattern used with mocking vault client.